paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexFile.java

```diff
@@ -0,0 +1,336 @@
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/** File index file format. Put all column and offset in the header. */
+public final class FileIndexFile {
+
+    private static final long MAGIC = 1493475289347502L;
+
+    enum Version {
+        V_1(1);
+
+        private final int version;
+
+        Version(int version) {
+            this.version = version;
+        }
+
+        public int version() {
+            return version;
+        }
+    }
+
+    public static Writer createWriter(OutputStream outputStream) {
+        return new Writer(outputStream);
+    }
+
+    public static Reader createReader(SeekableInputStream inputStream, RowType fileRowType) {
+        return new Reader(inputStream, fileRowType);
+    }
+
+    /** Writer for file index file. */
+    public static class Writer implements Closeable {
+
+        private final DataOutputStream dataOutputStream;
+
+        // for version compatible
+        private static final int REDUNDANT_LENGTH = 0;
+
+        public Writer(OutputStream outputStream) {
+            this.dataOutputStream = new DataOutputStream(outputStream);
+        }
+
+        public void writeColumnIndex(String indexType, Map<String, byte[]> bytesMap)
+                throws IOException {
+
+            Map<String, Pair<Integer, Integer>> bodyInfo = new HashMap<>();
+            byte[] body = constructBody(bytesMap, bodyInfo);
+
+            writeHead(indexType, bodyInfo);
+
+            // writeBody
+            dataOutputStream.write(body);
+        }
+
+        private byte[] constructBody(
+                Map<String, byte[]> bytesMap, Map<String, Pair<Integer, Integer>> bodyInfo)
+                throws IOException {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            for (Map.Entry<String, byte[]> entry : bytesMap.entrySet()) {
+                Integer startPosition = baos.size();
+                baos.write(entry.getValue());
+                bodyInfo.put(entry.getKey(), Pair.of(startPosition, baos.size() - startPosition));
+            }
+
+            return baos.toByteArray();
+        }
+
+        private void writeHead(String indexType, Map<String, Pair<Integer, Integer>> bodyInfo)
+                throws IOException {
+
+            int headLength = calculateHeadLength(indexType, bodyInfo);
+
+            // writeMagic
+            dataOutputStream.writeLong(MAGIC);
+            // writeVersion
+            dataOutputStream.writeInt(Version.V_1.version());
+            // writeHeadLength
+            dataOutputStream.writeInt(headLength);
+            // writeIndexType
+            dataOutputStream.writeUTF(indexType);
+            // writeColumnSize
+            dataOutputStream.writeInt(bodyInfo.size());
+            // writeColumnInfo, offset = headLength
+            for (Map.Entry<String, Pair<Integer, Integer>> entry : bodyInfo.entrySet()) {
```

Comment:
can we sort the key when write and use an algorithm (e.g. binary search) to speed up the search when read
in case of the large number of keys

Reply:
We `must` read all head columns and put them into a hashmap not a array list, so we don't need to sort.